### PR TITLE
Patch 283

### PR DESCRIPTION
**Pull request by saintentropy**
### Purpose

This patch address issues in [PSIRT-283](https://jira.autodesk.com/browse/PSIRT-283)

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @gregmarr @Dewb @QilongTang

### CODE REVIEW

File: src/DynamoUtilities/CertificateVerification.cs [hunk 1]
```diff
@@ -25,37 +303,19 @@ public static bool CheckAssemblyForValidCertificate(string assemblyPath)
...
-                cert = X509Certificate.CreateFromSignedFile(assemblyPath);
+                var validCert = WinTrustInterop.WinTrust.IsTrusted(assemblyPath);
+                if (validCert)
+                {
+                    return true;
+                }
             }
             catch
             {
                 throw new Exception(String.Format(
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: would it make sense to throw a more specific exception type here and test for this type explicitly in the tests?

File: src/DynamoUtilities/CertificateVerification.cs [hunk 2]
```diff
@@ -3,12 +3,290 @@
...
+
+            public static bool IsTrusted(string fileName)
+            {
+                return WinVerifyTrust(fileName) == 0;
+            }
+        }
+
+        private struct WINTRUST_FILE_INFO : IDisposable
+        {
+            public WINTRUST_FILE_INFO(string fileName, Guid subject)
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: /// comments I think would be useful

File: src/DynamoUtilities/CertificateVerification.cs [hunk 3]
```diff
@@ -25,37 +303,19 @@ public static bool CheckAssemblyForValidCertificate(string assemblyPath)
...
-            }
-
-            if (cert != null)
-            {
-                var cert2 = new System.Security.Cryptography.X509Certificates.X509Certificate2(cert);
-                if (cert2.Verify())
-                {
-                    return true;
-                }
+                    "A dll file found at {0} did not have a signed certificate.", assemblyPath));
```
Comment (gregmarr) [inline src/DynamoUtilities/CertificateVerification.cs]: Do you need to have this `throw` here?  Could you just let it fall through to the `throw` on the next line?

File: src/DynamoUtilities/CertificateVerification.cs [hunk 4]
```diff
@@ -3,12 +3,290 @@
...
+
+                fdwRevocationChecks = RevocationCheckFlags.None;
+
+                dwStateAction = StateAction.Ignore;
+
+                hWVTStateData = IntPtr.Zero;
+
+                pwszURLReference = IntPtr.Zero;
+
+                dwProvFlags = TrustProviderFlags.Safer;
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: I don't think we should use this.

File: src/DynamoUtilities/CertificateVerification.cs (hunk 3 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: The updated exceptions provide diffraction if the check failed due to exception or was actually false

File: src/DynamoUtilities/CertificateVerification.cs [hunk 5]
```diff
@@ -3,12 +3,194 @@
...
+            Execute = 0,
+            Install = 1
+        }
+        #endregion
+
+        #region WinTrust structures
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        class WinTrustFileInfo
+        {
+            UInt32 StructSize = (UInt32)Marshal.SizeOf(typeof(WinTrustFileInfo));
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: is this used anywhere?

File: src/DynamoUtilities/CertificateVerification.cs [hunk 6]
```diff
@@ -3,12 +3,194 @@
...
+                    Marshal.FreeCoTaskMem(pszFilePath);
+                    pszFilePath = IntPtr.Zero;
+                }
+            }
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        class WinTrustData
+        {
+            UInt32 StructSize = (UInt32)Marshal.SizeOf(typeof(WinTrustData));
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: is this used anywhere?

File: src/DynamoUtilities/CertificateVerification.cs (hunk 5 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: It is not

File: src/DynamoUtilities/CertificateVerification.cs [hunk 7]
```diff
@@ -3,12 +3,194 @@
...
+                    return ret;
+                }
+                finally
+                {
+                    // free the locally-held unmanaged memory in the data structures
+                    if (winTrustFileInfo != null) winTrustFileInfo.Dispose();
+                    if (winTrustData != null) winTrustData.Dispose();
+                }
+            }
+            private WinTrust() { }
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: why is this here?

File: src/DynamoUtilities/CertificateVerification.cs [hunk 8]
```diff
@@ -3,12 +3,194 @@
...
+            SubjectCertExpired = 0x800B0101,        // CERT_E_EXPIRED - Signer's certificate was expired
+            SubjectCertificateRevoked = 0x800B010C,     // CERT_E_REVOKED Subject's certificate was revoked
+            UntrustedRoot = 0x800B0109          // CERT_E_UNTRUSTEDROOT - A certification chain processed correctly but terminated in a root certificate that is not trusted by the trust provider.
+        }
+
+        public class WinTrust
+        {
+            private static readonly IntPtr INVALID_HANDLE_VALUE = new IntPtr(-1);
+            // GUID of the action to perform
+            private const string WINTRUST_ACTION_GENERIC_VERIFY_V2 = "{00AAC56B-CD44-11d0-8CC2-00C04FC295EE}";
```
Comment (mjkkirschner) [inline src/DynamoUtilities/CertificateVerification.cs]: should this be static?

File: src/DynamoUtilities/CertificateVerification.cs (hunk 5 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: But it fails if it is not set

File: src/DynamoUtilities/CertificateVerification.cs (hunk 6 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: Same as above

File: src/DynamoUtilities/CertificateVerification.cs (hunk 8 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: const is set at compile time vs static readonly set at runtime

File: src/DynamoUtilities/CertificateVerification.cs (hunk 7 above)
Comment (saintentropy) [reply src/DynamoUtilities/CertificateVerification.cs]: This does not seem to be necessary

File: src/DynamoUtilities/CertificateVerification.cs (hunk 5 above)
Comment (mjkkirschner) [reply src/DynamoUtilities/CertificateVerification.cs]: okay, I assume this is because the the class needs this field to exist when passed into native code and the layout is sequential so it needs to be first.

Comment (mjkkirschner) [reply src/DynamoUtilities/CertificateVerification.cs]: yes, I guess these map to `cbStruct`

File: test/Libraries/DynamoUtilitiesTests/CertificateVerificationTests.cs [hunk 9]
```diff
@@ -24,5 +24,43 @@ public void CheckAssemblyForValidCertificateTest()
...
+            ); 
+        }
+
+        [Test]
+        [Category("UnitTests")]
+        public void CheckAssemblyForValidCertificateTest_TransferredCertificate()
+        {
+            var executingDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            var directory = new DirectoryInfo(executingDirectory);
+            var testFilePath = Path.Combine(directory.Parent.Parent.Parent.FullName, "test", "pkgs_signed", "Modfied Signed Package", "bin", "LegitAssemblyTransplanted.dll");
```
Comment (Dewb) [inline test/Libraries/DynamoUtilitiesTests/CertificateVerificationTests.cs]: typo: `Modfied` => `Modified`

File: src/DynamoUtilities/CertificateVerification.cs (hunk 7 above)
Comment (Dewb) [reply src/DynamoUtilities/CertificateVerification.cs]: The purpose is to keep other classes from being able to construct instances of this otherwise static class.
